### PR TITLE
ARROW-10720: [C++] Add Rescale support for BasicDecimal256

### PR DESCRIPTION
**Pull request by Bei-z**

### CODE REVIEW

**Review by MingyuZhong — CHANGES_REQUESTED** [review]
Can you search for "TODO" in my [initial PR](https://github.com/apache/arrow/pull/8190/files) and see whether you can add the missing tests now?

**Comment by Bei-z** [comment]
> Can you search for "TODO" in my [initial PR](https://github.com/apache/arrow/pull/8190/files) and see whether you can add the missing tests now?

Thank you for pointing out! Added those tests.

**Review by MingyuZhong — APPROVED** [review]

File: cpp/src/arrow/util/decimal_test.cc [hunk 1]
```diff
@@ -985,6 +986,45 @@ TEST(Decimal128Test, Divide) {
   }
 }
 
+TEST(Decimal128Test, Rescale) {
+  ASSERT_EQ(Decimal128(11100), Decimal128(111).Rescale(0, 2).ValueOrDie());
```
Comment (pitrou) [inline cpp/src/arrow/util/decimal_test.cc:990]: Nit: we've got `ASSERT_OK_AND_EQ` for this. For example:
```c++
ASSERT_OK_AND_EQ(Decimal128(11100), Decimal128(111).Rescale(0, 2));
```

File: cpp/src/arrow/util/decimal_test.cc [hunk 2]
```diff
@@ -985,6 +986,45 @@ TEST(Decimal128Test, Divide) {
...
+  ASSERT_EQ(Decimal128(111), Decimal128(11100).Rescale(2, 0).ValueOrDie());
+  ASSERT_EQ(Decimal128(5), Decimal128(500000).Rescale(6, 1).ValueOrDie());
+  ASSERT_EQ(Decimal128(500000), Decimal128(5).Rescale(1, 6).ValueOrDie());
+  ASSERT_TRUE(Decimal128(555555).Rescale(6, 1).status().IsInvalid());
```
Comment (pitrou) [inline cpp/src/arrow/util/decimal_test.cc:994]:
```c++
ASSERT_RAISES(Invalid, Decimal128(555555).Rescale(6, 1));
```

File: cpp/src/arrow/util/decimal_test.cc [hunk 3]
```diff
@@ -985,6 +986,45 @@ TEST(Decimal128Test, Divide) {
...
+
+  // Test some random numbers.
+  Decimal128 multiplier(1);
+  for (int delta_scale = 0; delta_scale < 29;
+       delta_scale++, multiplier *= Decimal128(10)) {
```
Comment (pitrou) [inline cpp/src/arrow/util/decimal_test.cc:999]: You can limit the cost of this test by making the `delta_scale` loop the inner loop (generating random numbers is slightly expensive).

**Review by pitrou — COMMENTED** [review]
Nice! Just a couple small things.

File: cpp/src/arrow/util/decimal_test.cc (hunk 1 above)
Comment (Bei-z) [reply cpp/src/arrow/util/decimal_test.cc]: Changed accordingly. Thank you!

File: cpp/src/arrow/util/decimal_test.cc (hunk 2 above)
Comment (Bei-z) [reply cpp/src/arrow/util/decimal_test.cc]: Done. Thank you for pointing out!

File: cpp/src/arrow/util/decimal_test.cc (hunk 3 above)
Comment (Bei-z) [reply cpp/src/arrow/util/decimal_test.cc]: Changed accordingly. Thank you!

**Comment by pitrou** [comment]
CI failures should be unrelated, merging.

**Pull request closed without merging**